## 今天的作業
在鐵達尼資料集中，今天我們專注觀察變數之間的相關性，以Titanic_train.csv 中，首先將有遺失值的數值刪除，我們取 Titanic_train.csv 的年齡資料，試著將課堂中所學的方法應用上去。

* Q1: 產生一個新的變數(Age_above65_) Age>=65為 'Y'，其餘為'N'。  
* Q2: 添加女性和男性，產生一個新的變數(Age_above65_female)，女性或Age>=65為'Y'，其餘為'N'。  
*  Q3: 透過昨天課程的內容，驗證產生的兩個新變數，哪一個和目標變數(Survived_cate)的相關性較高?

In [1]:
# library
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import stats
import math
import statistics
import seaborn as sns
from IPython.display import display
import sklearn
print(sklearn.__version__)
#如果只有 0.19 記得要更新至 最新版本
%matplotlib inline


#顯示所有行
pd.set_option('display.max_columns', None)
#顯示所有列
#pd.set_option('display.max_rows', None)
#顯示行設定
pd.set_option('max_colwidth',100)

import pingouin as pg
import researchpy   
 

0.24.1


## 讀入資料

In [2]:
df_train = pd.read_csv("Titanic_train.csv")
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


In [3]:
#1.產稱一個新的變數 Survived_cate ，資料型態傳換成類別型態      
#2.把題目中的 Survived 用 Survived_cate 來做分析    
df_train['Survived_cate']=df_train['Survived']
df_train['Survived_cate']=df_train['Survived_cate'].astype('object')
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PassengerId    891 non-null    int64  
 1   Survived       891 non-null    int64  
 2   Pclass         891 non-null    int64  
 3   Name           891 non-null    object 
 4   Sex            891 non-null    object 
 5   Age            714 non-null    float64
 6   SibSp          891 non-null    int64  
 7   Parch          891 non-null    int64  
 8   Ticket         891 non-null    object 
 9   Fare           891 non-null    float64
 10  Cabin          204 non-null    object 
 11  Embarked       889 non-null    object 
 12  Survived_cate  891 non-null    object 
dtypes: float64(2), int64(5), object(6)
memory usage: 90.6+ KB
None


In [4]:
#我們先把遺失值刪除
## 取出資料後，把遺失值刪除
complete_data=df_train[['Age','Survived_cate','Sex']].dropna()
display(complete_data)

Age Survived_cate     Sex
0    22.0             0    male
1    38.0             1  female
2    26.0             1  female
3    35.0             1  female
4    35.0             0    male
..    ...           ...     ...
885  39.0             0  female
886  27.0             0    male
887  19.0             1  female
889  26.0             1    male
890  32.0             0    male

[714 rows x 3 columns]

### Q1: 產生一個新的變數(Age_above65_) Age>=65為 'Y'，其餘為'N'。


In [5]:
df_train["Age_above65_"] = df_train['Age'].apply((lambda x: 'Y' if x >=65 else 'N'))
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Survived_cate Age_above65_  
0      0         A/5 21171   7.2500   NaN        S             0            N  
1      0          PC 17599  71.2833   C85        C             1            N  
2      0  STON/O2. 3101282   7.9250   NaN        S             1            N  
3      0            113803  53.1000  C123        S             1            N  
4      0            373450   8.0500   NaN        S             0            N

### Q2: 添加女性和男性，產生一個新的變數(Age_above65_female)，女性或Age>=65為'Y'，其餘為'N'。
* 暗示: 觀看下面影片找答案，https://www.youtube.com/watch?v=X2d-wUt5azY

In [6]:
def age_abv65_female_map(row):
    if row.Sex == "female" or row.Age >=65:
        return "Y"
    else:
        return "N"
df_train["Age_above65_female"] = df_train.apply(age_abv65_female_map, axis=1)
df_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                    Name     Sex   Age  SibSp  \
0                                Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                                 Heikkinen, Miss. Laina  female  26.0      0   
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                               Allen, Mr. William Henry    male  35.0      0   
..                                                   ...     ...   ...    ...   
886                                Montvila, Rev. Juozas    male  27.0      0   
887                         Graham, Miss. Margaret Edith  female  19.0      0   
888             Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                                Behr, Mr. Karl Howell    male  26.0      0   
890                                  Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Survived_cate  \
0        0         A/5 21171   7.2500   NaN        S             0   
1        0          PC 17599  71.2833   C85        C             1   
2        0  STON/O2. 3101282   7.9250   NaN        S             1   
3        0            113803  53.1000  C123        S             1   
4        0            373450   8.0500   NaN        S             0   
..     ...               ...      ...   ...      ...           ...   
886      0            211536  13.0000   NaN        S             0   
887      0            112053  30.0000   B42        S             1   
888      2        W./C. 6607  23.4500   NaN        S             0   
889      0            111369  30.0000  C148        C             1   
890      0            370376   7.7500   NaN        Q             0   

    Age_above65_ Age_above65_female  
0              N                  N  
1              N                  Y  
2              N                  Y  
3              N                  Y  
4              N                  N  
..           ...                ...  
886            N                  N  
887            N                  Y  
888            N                  Y  
889            N                  N  
890            N                  N  

[891 rows x 15 columns]

### Q3: 透過昨天課程的內容，驗證產生的兩個新變數，哪一個和目標變數(Survived_cate)的相關性較高?
* 提示:    
  首先觀察一下這些變數的資料型態後，再來想要以哪一種判斷倆倆的相關性。     

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df_train["Age_above65_YN"] = le.fit_transform(df_train['Age_above65_'])
le = LabelEncoder()
df_train["Age_above65_female_YN"] = le.fit_transform(df_train['Age_above65_female'])
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Survived_cate Age_above65_  \
0      0         A/5 21171   7.2500   NaN        S             0            N   
1      0          PC 17599  71.2833   C85        C             1            N   
2      0  STON/O2. 3101282   7.9250   NaN        S             1            N   
3      0            113803  53.1000  C123        S             1            N   
4      0            373450   8.0500   NaN        S             0            N   

  Age_above65_female  Age_above65_YN  Age_above65_female_YN  
0                  N               0                      0  
1                  Y               0                      1  
2                  Y               0                      1  
3                  Y               0                      1  
4                  N               0                      0

In [8]:
df_train.corr()[['Age_above65_YN','Age_above65_female_YN']]

Age_above65_YN  Age_above65_female_YN
PassengerId                 -0.018533              -0.046866
Survived                    -0.067344               0.523793
Pclass                      -0.077787              -0.148747
Age                          0.345319              -0.004899
SibSp                       -0.043834               0.103709
Parch                       -0.027725               0.237272
Fare                        -0.007425               0.179250
Age_above65_YN               1.000000               0.147544
Age_above65_female_YN        0.147544               1.000000

### Age_above65_female 相關度為 0.523793 大於 Age_above65 相關度 0.067344

In [9]:
def judgment_CramerV(df,V):
    if df == 1:
        if V < 0.10:
            qual = 'negligible'
        elif V < 0.30:
            qual = 'small'
        elif V < 0.50:
            qual = 'medium'
        else:
            qual = 'large'
    elif df == 2:
        if V < 0.07:
            qual = 'negligible'
        elif V < 0.21:
            qual = 'small'
        elif V < 0.35:
            qual = 'medium'
        else:
            qual = 'large'
    elif df == 3:
        if V < 0.06:
            qual = 'negligible'
        elif V < 0.17:
            qual = 'small'
        elif V < 0.29:
            qual = 'medium'
        else:
            qual = 'large'
    elif df == 4:
        if V < 0.05:
            qual = 'negligible'
        elif V < 0.15:
            qual = 'small'
        elif V < 0.25:
            qual = 'medium'
        else:
            qual = 'large'
    else:
        if V < 0.05:
            qual = 'negligible'
        elif V < 0.13:
            qual = 'small'
        elif V < 0.22:
            qual = 'medium'
        else:
            qual = 'large'
    return(qual)

In [10]:
crosstab, res = researchpy.crosstab(df_train['Age_above65_'], df_train['Survived_cate'], test='chi-square')
print("Cramer's value is",res.loc[2,'results'])
contTable = pd.crosstab(df_train['Age_above65_'], df_train['Survived_cate'])
df = min(contTable.shape[0], contTable.shape[1]) - 1
print(judgment_CramerV(df,res.loc[2,'results']))

Cramer's value is 0.0673
negligible


In [11]:
crosstab, res = researchpy.crosstab(df_train['Age_above65_female'], df_train['Survived_cate'], test='chi-square')
print("Cramer's value is",res.loc[2,'results'])
contTable = pd.crosstab(df_train['Age_above65_female'], df_train['Survived_cate'])
df = min(contTable.shape[0], contTable.shape[1]) - 1
print(judgment_CramerV(df,res.loc[2,'results']))

Cramer's value is 0.5238
large


### Age_above65_female 相關度為 0.5238 大於 Age_above65 相關度 0.0673